Initial review and redo of screenshots (#282)
carpentries-incubator/python-intermediate-development: _episodes/41-code-review.md

   and then submit the pull request by clicking `Create pull request` button (in the new window).
![Submitting a pull request.](../fig/github-submit-pull-request.png){: .image-with-shadow width="900px"}
7. At this point, the code review process is initiated.

We will now discuss what to look for in a code review,
before practising it on this fictional change.

### Reviewing a Pull Request

Once a review has been raised it is over to the reviewer to review the code
and submit feedback.

Reviewing code effectively takes practice.
However, here is some guidance on what you should
be looking for when reviewing a piece of code.

#### Things to Look for in a Code Review

Start by understanding what the code _should_ do, by reading the specification/user requirements,
the pull request description and talking to the developer.
In this case, understand what [SR1.1.1](../31-software-requirements/index.html#solution-requirements) means.

Once you're happy, start reading the code (skip the test code for now - we will come back to it later). 
You're going to be assessing the code in the following key areas.

##### Is the proposed code readable?

> * Think about do the names of the variables, do they [follow guidelines for good

Reviewer: ```suggestion
* Think about the names of the variables, do they [follow guidelines for good
```

names?](../15-coding-conventions/index.html#l#naming-conventions)
* Do you understand what conditions in each if statements are for?

> * Do the function names match the behavior of the function.

Reviewer: ```suggestion
* Do the function names match the behavior of the function?
```

##### Is the proposed code a minimal change?

* Does the code reimplement anything that already exists, either
elsewhere in the codebase or in a library you know about?
* Does the code implement something that isn't on the ticket?

> ##### Is the structure of the proposed code clear?

Reviewer: ```suggestion
##### Is the structure of the code clear?
```

I'd remove 'proposed' here, since you're also looking at the change in the context of the rest of the code as well, i.e. that the rest of the code's structure isn't negatively affected by the proposed change.

* Are there clear and useful comments that explain complex designs
and focus on the "why/because" rather than the "what/how"?

#### Things Not to Look for in a Code Review

The overriding priority for reviewing code should be making sure progress is being made -

> don't let perfect be the enemy of good here.

Reviewer: ```suggestion
don't let perfect be the enemy of the good here.
```

According to [“Best Kept Secrets of Peer Code Review” (Cohen, 2006)](https://www.amazon.co.uk/Best-Kept-Secrets-Peer-Review/dp/1599160676)
the first hour of reviewing code is the most effective, with diminishing returns after that.

To that end, here are a few things you *shouldn't* be trying to spot when reviewing:
* Linting issues, or anything else that an automated tool can spot - get the Continuous Integration (CI) to do it.
* Bugs - instead make sure there are tests for all cases.
* Issues that pre-date the change - raise separate PRs fixing these issues separately to avoid heading down a rabbit hole.
* Architecture re-writes - try to have design discussions upfront, 
or else have a meeting to decide whether the code needs to be rewritten.

#### Adding a review comment

> 1. To be able to add your review to a pull request created on your course mate's repository, 

Reviewer: ```suggestion
1. To be able to add your review to a pull request created on your fellow learner's repository, 
```

Again, for consistency.

> 1. To be able to add your review to a pull request created on your course mate's repository, 

Reviewer: I recall we discussed a callout about the general case of code review where you don't strictly need to add people to your repo to do a code review, which should be added.

Author: Looks like I failed to commit but did a push which did not have any affect on my latest changes - sorry @steve-crouch

they will need to add you as a collaborator on their repository first. They can do that from 
the `Settings` tab of the repository, then `Collaborators and teams` tab on the left, 
then clicking the `Add people` button. 
Once they find you by your GitHub name - you will receive an invitation via email to join the 
repository as a collaborator. 
You will have to do the same for the collaborator doing the review on your repository.
![Adding a collaborator in GitHub](../fig/github-add-collaborator.png){: .image-with-shadow width="900px"}
2. Open up the pull request from the GitHub's `Pull Requests` tab on the home page 

> of your course mate's software repository, then head to the `Files changed` tab.

Reviewer: ```suggestion
of your fellow learner's software repository, then head to the `Files changed` tab.
```

![Adding a review comment to a pull request](../fig/github-pull-request-add-comment.png){: .image-with-shadow width="800px"}
You can also add comments referring to multiple lines by clicking the plus and
dragging down over the relevant lines.
If you want to make a concrete suggestion or a change to the code directly, 
such as renaming a variable, you can click the `Add a suggestion` button 
(which looks like a document with a plus and a minus in it).
This will populate the comment with the existing code, and you can edit it to be
what you think the code should be.
![Adding a suggestion to a pull request](../fig/github-pull-request-add-suggestion.png){: .image-with-shadow width="800px"} 
GitHub will then provide a button for the code author to apply your change directly.
4. Write your comment in the box, and then click `Start review`.
This will save your comment, but not publish it yet.
You can use `Add single comment` button to immediately post a comment.
However, it is best to batch the comments into a single review, so that the author
knows when you have finished adding comments
(and avoid spamming their email with notifications).

> 5. Continue adding comments in this way, using the `Add review comment` button

Reviewer: Looking at the original PR, this is supposed to be done in a group in the following exercise, so probably should be removed.

Author: Think it is OK to keep it as is as reviewers may have more comments to add here, before we move to the next exercise and they even more comments around tests @steve-crouch

Author: Improved via f51ee2e2639cb1540aec389f6ab060389c49c2a2

on subsequent comments.

> ## Effective Review Comments
>
> * Make sure your review comments are specific and actionable.

> > * Try to be as specific as you can - instead "this code is unclear"

Reviewer: ```suggestion
> * Try to be as specific as you can - instead of "this code is unclear"
```

> > say "I don't understand what values this variable can hold".

Reviewer: ```suggestion
> instead say "I don't understand what values this variable can hold".
```

![Adding a review comment to a pull request](../fig/github-pull-request-add-comment.png){: .image-with-shadow width="800px"}
You can also add comments referring to multiple lines by clicking the plus and
dragging down over the relevant lines.
If you want to make a concrete suggestion or a change to the code directly, 
such as renaming a variable, you can click the `Add a suggestion` button 
(which looks like a document with a plus and a minus in it).
This will populate the comment with the existing code, and you can edit it to be
what you think the code should be.
![Adding a suggestion to a pull request](../fig/github-pull-request-add-suggestion.png){: .image-with-shadow width="800px"} 
GitHub will then provide a button for the code author to apply your change directly.
4. Write your comment in the box, and then click `Start review`.
This will save your comment, but not publish it yet.
You can use `Add single comment` button to immediately post a comment.
However, it is best to batch the comments into a single review, so that the author
knows when you have finished adding comments
(and avoid spamming their email with notifications).

> 5. Continue adding comments in this way, using the `Add review comment` button

Author: ```suggestion
5. Continue adding comments in this way, if you have any, using the `Add review comment` button.
```
